
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Preprocessing &#8212; plco-analysis v1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="&lt;no title&gt;" href="Primary%20Analysis.html" />
    <link rel="prev" title="&lt;no title&gt;" href="Installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="preprocessing">
<h1>Preprocessing<a class="headerlink" href="#preprocessing" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This content should provide an overview of the one-time steps you’ll need to perform when running this pipeline for the first time
on a given chip or imputation build. You do not need to run this multiple times for the same build! This should save you lots of time.</p>
<p>:: <span class="target" id="relatedness-pipeline">relatedness pipeline</span>:
Relatedness
~~~~~~~~~~~</p>
<p>Relatedness estimation is primarily required as an intermediate in the ancestry estimation process using <a href="#id1"><span class="problematic" id="id2">graf_</span></a>. The files are stored
and can be used for other QC reasons of course; see <cite>relatedness/PLCO_{chip_name}.relatedness.txt</cite> for relevant output files in standard
graf format.</p>
<ul class="simple">
<li><p>Usage: <cite>make relatedness</cite></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p>chip freeze</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>input chip data are in <a href="#id3"><span class="problematic" id="id4">plink_</span></a> bed/bim/fam format</p></li>
<li><p>variant annotations are in GRCh38</p></li>
</ul>
</li>
</ul>
<p>:: _graf: <a class="reference external" href="https://github.com/ncbi/graf">https://github.com/ncbi/graf</a>
:: _plink: <a class="reference external" href="https://www.cog-genomics.org/plink/">https://www.cog-genomics.org/plink/</a></p>
<p>:: <span class="target" id="ancestry-pipeline">ancestry pipeline</span>:
Ancestry Estimation
~~~~~~~~~~~~~~~~~~~</p>
<p>Ancestry estimation is required for chip processing and various sanity checks. As with relatedness, this is computed with <a href="#id5"><span class="problematic" id="id6">graf_</span></a>. The files
are stored and can be used for various QC purposes; see <cite>ancestry/PLCO_{chip_name}.graf_estimates.txt</cite>.</p>
<p>Note that the final ancestry calls listed above are modified according to the consensus instructions of the “Atlas” analysis group.
Subjects from the default “African” <a href="#id7"><span class="problematic" id="id8">graf_</span></a> ancestry are merged into the “African American” label to more consistently represent
the sampling distribution of the PLCO project. Subjects from the default “Other Asian or Pacific Islander” <a href="#id9"><span class="problematic" id="id10">graf_</span></a> ancestry are merged
into the “East Asian” label according to the instructions of collaborators.</p>
<ul class="simple">
<li><p>Usage: <cite>make ancestry</cite></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p><a class="reference internal" href="#relatedness-pipeline">relatedness pipeline</a></p></li>
<li><p>chip freeze</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>same as <a class="reference internal" href="#relatedness-pipeline">relatedness pipeline</a></p></li>
</ul>
</li>
</ul>
<p>::<span class="target" id="cleaned-chip-by-ancestry-pipeline">cleaned chip by ancestry pipeline</span>:
Genotype Cleaning by Ancestry
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</p>
<p>The PLCO chip freeze provides five files: one for each platform (GSA, Oncoarray, OmniX, Omni25, Omni5), with each <a href="#id11"><span class="problematic" id="id12">graf_</span></a> ancestry group
combined into a single <a href="#id13"><span class="problematic" id="id14">plink_</span></a> bed/bim/fam dataset. These files need to be processed and cleaned with the following aims:</p>
<ul class="simple">
<li><p>GWAS chip cleaning must take place on a per-ancestry level. This pipeline does not assume this step has been correctly handled,
and will apply such cleaning as is minimally required for successful execution</p></li>
<li><p>Individual <a href="#id15"><span class="problematic" id="id16">plink_</span></a> files by ancestry/platform are required for certain downstream processing steps</p></li>
<li><p>PCA estimation by ancestry/platform are required for certain downstream association steps</p></li>
<li><p>An intermediate of PCA and IBS/IBD estimation, LD pruned genotypes by ancestry/platform, are required for certain downstream
processing steps</p></li>
</ul>
<p>Note that some of these files and calculations, such as <a class="reference internal" href="#heterozygosity-outliers">heterozygosity outliers</a> and <a class="reference internal" href="#ibs-ibd-estimates">IBS/IBD estimates</a>, are not directly
used by the pipeline in its current form but were useful in the processing of the PLCO chip freeze, and are included for posterity.</p>
<p>:: <span class="target" id="heterozygosity-outliers">heterozygosity outliers</span>: <a class="reference external" href="https://www.cog-genomics.org/plink/1.9/basic_stats#ibc">https://www.cog-genomics.org/plink/1.9/basic_stats#ibc</a>
:: <span class="target" id="ibs-ibd-estimates">IBS/IBD estimates</span>: <a class="reference external" href="https://www.cog-genomics.org/plink/1.9/ibd">https://www.cog-genomics.org/plink/1.9/ibd</a></p>
<ul class="simple">
<li><p>Usage: <cite>make cleaned-chips-by-ancestry</cite></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p><a class="reference internal" href="#ancestry-pipeline">ancestry pipeline</a></p></li>
<li><p>chip freeze</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>same as <a href="#id17"><span class="problematic" id="id18">ancestry_</span></a> pipeline</p></li>
<li><p>this was developed separately from the rest of the pipeline and folded in later. thus there exist some redundancies
between this and other aspects of the pipeline that might be refactored out in later development. this is also one
of the oldest of the subpipelines, so structurally it differs from some of the other pipelines. nevertheless the
overall idea is the same: apply the same methods to each chip and ancestry in a project in a uniform and dynamic fashion</p></li>
<li><p>for PLCO, these internal QC steps have already been applied upstream of the latest chip freeze. thus it is expected
that the results for these platforms for heterozygosity are empty. if this is not true in later applications of the
pipeline, some amount of restructuring may be necessary</p></li>
<li><p>for PLCO, an entire cleaning process based on IBS/IBD pruning was applied upstream of the chip freeze. that process
(or frankly some better version of it) should be applied to chip data before actual use in association; but no such
method is implemented here. if there’s interest, this might be the place to do it, someday during future development</p></li>
</ul>
</li>
</ul>
<p>:: <span class="target" id="bgen-pipeline">bgen_pipeline</span>:
Imputed Data Reformatting
~~~~~~~~~~~~~~~~~~~~~~~~~</p>
<p>PLCO uses imputed data files from <a href="#id19"><span class="problematic" id="id20">minimac4_</span></a> from the <a class="reference internal" href="#michigan-imputation-server">Michigan Imputation Server</a>. These files come in a standard VCF +
annotation format; the pipeline further assumes that post-imputation (Rsq-based) QC has been applied to the files, and you’re
working with something like an imputed data freeze.</p>
<p>Association tools sometimes handle VCF format input, but often offer speed improvements when operating on other formats.
Thus, based on the tools requested during the initial “Atlas” testing phase, the format <a href="#id21"><span class="problematic" id="id22">bgen_</span></a> has been selected as a common
analysis format for all tools. Imputed data files for each platform/ancestry combination are generated once per dataset,
and then used as fixed input for all association testing.</p>
<p>:: _minimac4: <a class="reference external" href="https://genome.sph.umich.edu/wiki/Minimac4">https://genome.sph.umich.edu/wiki/Minimac4</a>
:: <span class="target" id="michigan-imputation-server">Michigan Imputation Server</span>: <a class="reference external" href="https://imputationserver.sph.umich.edu/index.html">https://imputationserver.sph.umich.edu/index.html</a>
:: _bgen: <a class="reference external" href="https://www.well.ox.ac.uk/~gav/bgen_format/">https://www.well.ox.ac.uk/~gav/bgen_format/</a>
:: _saige: <a class="reference external" href="https://github.com/weizhouUMICH/SAIGE">https://github.com/weizhouUMICH/SAIGE</a>
:: _bolt-lmm: <a class="reference external" href="https://alkesgroup.broadinstitute.org/BOLT-LMM/BOLT-LMM_manual.html">https://alkesgroup.broadinstitute.org/BOLT-LMM/BOLT-LMM_manual.html</a></p>
<ul class="simple">
<li><p>Usage: <cite>make bgen</cite></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p>imputed data freeze</p></li>
<li><p>list of unique subjects across platforms with corresponding platform to choose their unique instance from</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>currently <a href="#id23"><span class="problematic" id="id24">saige_</span></a> and <a href="#id25"><span class="problematic" id="id26">bolt-lmm_</span></a> use <a href="#id27"><span class="problematic" id="id28">bgen_</span></a> as an input format. if you’re not using those tools, or another that accepts
<a href="#id29"><span class="problematic" id="id30">bgen_</span></a> input, then there’s no reason to run this pipeline and waste a ton of hard drive space</p></li>
<li><p>the <a href="#id31"><span class="problematic" id="id32">bgen_</span></a> format in use here is v1.2, based on <a href="#id33"><span class="problematic" id="id34">bolt-lmm_</span></a> documentation and other software support suggesting that
that’s the most efficient version accepted by all current tools. this may need to be changed in the future</p></li>
<li><p>the <a href="#id35"><span class="problematic" id="id36">bgen_</span></a> reformatting process is conducted using <a href="#id37"><span class="problematic" id="id38">plink_</span></a>. the resultant <cite>*.sample</cite> files are slightly malformatted,
and so an additional step fixes the included <cite>NA</cite> values by setting them to <cite>0</cite>. this could well be changed in
a future version depending on upstream behavior</p></li>
</ul>
</li>
</ul>
<p>:: <span class="target" id="genomes-pipeline">1000 Genomes pipeline</span>:
1000 Genomes Reference Data
~~~~~~~~~~~~~~~~~~~~~~~~~~~</p>
<p>Certain pipelines (see below) use derived information from the <a class="reference internal" href="#genomes-project">1000 Genomes Project</a>. This is (always?) split
by supercontinent.</p>
<p>:: <span class="target" id="genomes-project">1000 Genomes Project</span>: <a class="reference external" href="https://www.internationalgenome.org">https://www.internationalgenome.org</a></p>
<ul class="simple">
<li><p>Usage: <cite>make 1KG_files</cite></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p>a functional internet connection</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>the 1000 Genomes files downloaded are frozen at a particular latest release according to the configuration information
in <cite>Makefile.config</cite>. that can obviously be changed if you want</p></li>
<li><p>most target installations should actually have some sort of copy of the 1000 Genomes data present already somewhere
on their filesystem; however, this pipeline is not designed to support that as-is. it should be pretty easy to modify
if you really want</p></li>
</ul>
</li>
</ul>
<p>:: <span class="target" id="ldsc-pipeline">ldsc pipeline</span>:
LD Score/BOLT Reference Files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</p>
<p><a href="#id39"><span class="problematic" id="id40">ldsc_</span></a> and <a href="#id41"><span class="problematic" id="id42">bolt-lmm_</span></a> use data files derived from 1000 Genomes reference supercontinental data. Though some standard
reference data are provided in these packages (at least at time of initial download), that information only
covered European subjects, and more generalized data were/are needed.</p>
<p>:: _ldsc: <a class="reference external" href="https://github.com/bulik/ldsc">https://github.com/bulik/ldsc</a></p>
<ul class="simple">
<li><p>Usage: <cite>make ldsc</cite></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p><a class="reference internal" href="#genomes-pipeline">1000 Genomes pipeline</a></p></li>
<li><p>1000 Genomes supercontinental definitions (included by default)</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>the reference files included with <a href="#id43"><span class="problematic" id="id44">bolt-lmm_</span></a> have a description of (in theory) how they were generated, but when you
actually run <a href="#id45"><span class="problematic" id="id46">ldsc_</span></a> you find that none of the output matches the included reference files exactly. this is probably
some sort of weird versioning issue. regardless, this pipeline just hacks the result into submission. that results
in some discrepancies from the stock reference files, but there’s no indication of exactly which subjects/variants
were used for those files, so that’s not unexpected. basically: ymmv</p></li>
<li><p>default built-in files include an African American (<cite>AFRAMR</cite>) meta-group for appropriate subjects. note however
that “African American” as a human genetics group label is a very heterogeneous group, so there’s no guarantee
that this reference group will be appropriate for a given set of African American study subjects</p></li>
</ul>
</li>
</ul>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">plco-analysis</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Preprocessing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="Installation.html" title="previous chapter">&lt;no title&gt;</a></li>
      <li>Next: <a href="Primary%20Analysis.html" title="next chapter">&lt;no title&gt;</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Cameron Palmer.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.4.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/Preprocessing.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>