
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Preprocessing &#8212; plco-analysis v1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Primary Analysis" href="Primary%20Analysis.html" />
    <link rel="prev" title="Installation" href="Installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="preprocessing">
<h1>Preprocessing<a class="headerlink" href="#preprocessing" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This content should provide an overview of the one-time steps you’ll need to perform when running this pipeline for the first time
on a given chip or imputation build. You do not need to run this multiple times for the same build! This should save you lots of time.</p>
</div>
<div class="section" id="relatedness">
<span id="relatedness-pipeline"></span><h2>Relatedness<a class="headerlink" href="#relatedness" title="Permalink to this headline">¶</a></h2>
<p>Relatedness estimation is primarily required as an intermediate in the ancestry estimation process using <a class="reference external" href="https://github.com/ncbi/graf">graf</a>. The files are stored
and can be used for other QC reasons of course; see <strong>relatedness/PLCO_{chip_name}.relatedness.txt</strong> for relevant output files in standard
graf format.</p>
<ul class="simple">
<li><p>Usage: <strong>make relatedness</strong></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p>chip freeze</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>input chip data are in <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a> bed/bim/fam format</p></li>
<li><p>variant annotations are in GRCh38</p></li>
</ul>
</li>
</ul>
<div class="topic">
<p class="topic-title">Debugging</p>
<p>Very little can go wrong in this early step of the pipeline. Most likely any issues would be related to fundamental
misunderstandings about the formatting of the input chip data.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><a class="reference external" href="https://github.com/ncbi/graf">graf</a> is mildly frustrating to use: it has some non-compliant behaviors. Notably, its exit codes are not standard,
so it doesn’t exit <cite>0</cite> on success. Always check the output log from <a class="reference external" href="https://github.com/ncbi/graf">graf</a> before proceeding! And note that the overall
<cite>make</cite> run will have some “exit code ignored” warnings due to this behavior.</p>
<p>Additionally, <a class="reference external" href="https://github.com/ncbi/graf">graf</a> complains when the output files it tries to create already exist. So, if you’re running <a class="reference external" href="https://github.com/ncbi/graf">graf</a> in an
existing directory, you will likely need to purge intermediates (or just kill the “relatedness/” directory and check it out again)
before rerunning. However, this is an early step that doesn’t expect to be rerun frequently. It can definitely be patched to
work better; or you can use the version in the upstream QC pipeline that’s much better.</p>
</div>
</div>
<div class="section" id="ancestry-estimation">
<span id="ancestry-pipeline"></span><h2>Ancestry Estimation<a class="headerlink" href="#ancestry-estimation" title="Permalink to this headline">¶</a></h2>
<p>Ancestry estimation is required for chip processing and various sanity checks. As with relatedness, this is computed with <a class="reference external" href="https://github.com/ncbi/graf">graf</a>. The files
are stored and can be used for various QC purposes; see <strong>ancestry/PLCO_{chip_name}.graf_estimates.txt</strong>.</p>
<p>Note that the final ancestry calls listed above are modified according to the consensus instructions of the “Atlas” analysis group.
Subjects from the default “African” <a class="reference external" href="https://github.com/ncbi/graf">graf</a> ancestry are merged into the “African American” label to more consistently represent
the sampling distribution of the PLCO project. Subjects from the default “Other Asian or Pacific Islander” <a class="reference external" href="https://github.com/ncbi/graf">graf</a> ancestry are merged
into the “East Asian” label according to the instructions of collaborators.</p>
<ul class="simple">
<li><p>Usage: <strong>make ancestry</strong></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p><a class="reference internal" href="#relatedness-pipeline">relatedness pipeline</a></p></li>
<li><p>chip freeze</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>same as <a class="reference internal" href="#relatedness-pipeline">relatedness pipeline</a></p></li>
</ul>
</li>
</ul>
<div class="topic">
<p class="topic-title">Debugging</p>
<p>As with the <a class="reference internal" href="#relatedness-pipeline">relatedness pipeline</a>, there isn’t much that can go particularly wrong here. If you find anything, do feel free
to add it to this note block for posterity.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>See the <a class="reference internal" href="#relatedness-pipeline">relatedness pipeline</a> for warnings about <a class="reference external" href="https://github.com/ncbi/graf">graf</a> idiosyncracies. There’s an additional issue with
<a class="reference external" href="https://github.com/ncbi/graf">graf</a> at this step, as it uses a mildly malformed Perl script. The <a class="reference external" href="https://docs.conda.io/en/latest/">conda</a> installed version of <a class="reference external" href="https://github.com/ncbi/graf">graf</a> from the <a class="reference external" href="https://github.com/NCI-CGR/conda-cgr">CGR conda channel</a>
has an installation hack that repairs one of the issues; but I think it’s likely there are more corner case issues that may
yet pop up.</p>
</div>
</div>
<div class="section" id="genotype-cleaning-by-ancestry">
<span id="cleaned-chip-by-ancestry-pipeline"></span><h2>Genotype Cleaning by Ancestry<a class="headerlink" href="#genotype-cleaning-by-ancestry" title="Permalink to this headline">¶</a></h2>
<p>The PLCO chip freeze provides five files: one for each platform (GSA, Oncoarray, OmniX, Omni25, Omni5), with each <a class="reference external" href="https://github.com/ncbi/graf">graf</a> ancestry group
combined into a single <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a> bed/bim/fam dataset. These files need to be processed and cleaned with the following aims:</p>
<ul class="simple">
<li><p>GWAS chip cleaning must take place on a per-ancestry level. This pipeline does not assume this step has been correctly handled,
and will apply such cleaning as is minimally required for successful execution</p></li>
<li><p>Individual <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a> files by ancestry/platform are required for certain downstream processing steps</p></li>
<li><p>PCA estimation by ancestry/platform are required for certain downstream association steps</p></li>
<li><p>An intermediate of PCA and IBS/IBD estimation, LD pruned genotypes by ancestry/platform, are required for certain downstream
processing steps</p></li>
</ul>
<p>Note that some of these files and calculations, such as <a class="reference external" href="https://www.cog-genomics.org/plink/1.9/basic_stats#ibc">heterozygosity outliers</a> and <a class="reference external" href="https://www.cog-genomics.org/plink/1.9/ibd">IBS/IBD estimates</a>, are not directly
used by the pipeline in its current form but were useful in the processing of the PLCO chip freeze, and are included for posterity.</p>
<ul class="simple">
<li><p>Usage: <strong>make cleaned-chips-by-ancestry</strong></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p><a class="reference internal" href="#ancestry-pipeline">ancestry pipeline</a></p></li>
<li><p>chip freeze</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>same as <a class="reference internal" href="#ancestry-pipeline">ancestry pipeline</a></p></li>
<li><p>this was developed separately from the rest of the pipeline and folded in later. thus there exist some redundancies
between this and other aspects of the pipeline that might be refactored out in later development. this is also one
of the oldest of the subpipelines, so structurally it differs from some of the other pipelines. nevertheless the
overall idea is the same: apply the same methods to each chip and ancestry in a project in a uniform and dynamic fashion</p></li>
<li><p>for PLCO, these internal QC steps have already been applied upstream of the latest chip freeze. thus it is expected
that the results for these platforms for heterozygosity are empty. if this is not true in later applications of the
pipeline, some amount of restructuring may be necessary</p></li>
<li><p>for PLCO, an entire cleaning process based on IBS/IBD pruning was applied upstream of the chip freeze. that process
(or frankly some better version of it) should be applied to chip data before actual use in association; but no such
method is implemented here. if there’s interest, this might be the place to do it, someday during future development</p></li>
</ul>
</li>
</ul>
<div class="topic">
<p class="topic-title">Debugging</p>
<p>This pipeline extensively uses <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a> for filtering and QC operations. <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a>’s memory allocation is limited to 16G
in <strong>Makefile.config</strong>. That’s a completely <em>ad hoc</em> bit of nonsense that may need to be changed depending on your
individual project’s parameters.</p>
<p>The pipeline is designed to allow different combinations of platform/ancestry to not exist. That seems to work well,
but some issues may pop up if <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a> finds something it doesn’t like in a small dataset.</p>
<p>The IBS/IBD calculation with <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a> <strong>–genome</strong> is somewhat quirkly set up. For datasets above a fixed (configurable)
threshold of number of subjects, the IBS/IBD calculation is split into chunks with <strong>–parallel</strong> and then glued back
together in a separate rule. These various thresholds were selected to make PLCO/GSA/Europeans run reasonably efficiently.
For much larger chips, you may need to fiddle with the thresholds and number of quasiparallelized jobs to make things
go ok.</p>
<p>PCA with <a class="reference external" href="https://alkesgroup.broadinstitute.org/EIGENSOFT/">smartpca</a> makes use of the frankly underdocumented fastmode to make things go in a reasonable amount of time.
This is not ideal in a variety of circumstances, most notably if you want to do PCA outlier removal. PCA in this context
is really a blunt instrument, and not suitable for chip QC, which should have been handled earlier. So just… keep that
in mind. For posterity, I’ll record that the eigenvalues from <a class="reference external" href="https://alkesgroup.broadinstitute.org/EIGENSOFT/">smartpca</a> fastmode are included in the header of the raw
<a class="reference external" href="https://alkesgroup.broadinstitute.org/EIGENSOFT/">smartpca</a> output.</p>
</div>
</div>
<div class="section" id="imputed-data-reformatting">
<span id="bgen-pipeline"></span><h2>Imputed Data Reformatting<a class="headerlink" href="#imputed-data-reformatting" title="Permalink to this headline">¶</a></h2>
<p>PLCO uses imputed data files from <a class="reference external" href="https://genome.sph.umich.edu/wiki/Minimac4">minimac4</a> from the <a class="reference external" href="https://imputationserver.sph.umich.edu/index.html">Michigan Imputation Server</a>. These files come in a standard VCF +
annotation format; the pipeline further assumes that post-imputation (Rsq-based) QC has been applied to the files, and you’re
working with something like an imputed data freeze.</p>
<p>Association tools sometimes handle VCF format input, but often offer speed improvements when operating on other formats.
Thus, based on the tools requested during the initial “Atlas” testing phase, the format <a class="reference external" href="https://www.well.ox.ac.uk/~gav/bgen_format/">bgen</a> has been selected as a common
analysis format for all tools. Imputed data files for each platform/ancestry combination are generated once per dataset,
and then used as fixed input for all association testing.</p>
<ul class="simple">
<li><p>Usage: <strong>make bgen</strong></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p>imputed data freeze</p></li>
<li><p>list of unique subjects across platforms with corresponding platform to choose their unique instance from</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>currently <a class="reference external" href="https://github.com/weizhouUMICH/SAIGE">saige</a> and <a class="reference external" href="https://alkesgroup.broadinstitute.org/BOLT-LMM/BOLT-LMM_manual.html">bolt-lmm</a> use <a class="reference external" href="https://www.well.ox.ac.uk/~gav/bgen_format/">bgen</a> as an input format. if you’re not using those tools, or another that accepts
<a class="reference external" href="https://www.well.ox.ac.uk/~gav/bgen_format/">bgen</a> input, then there’s no reason to run this pipeline and waste a ton of hard drive space</p></li>
<li><p>the <a class="reference external" href="https://www.well.ox.ac.uk/~gav/bgen_format/">bgen</a> format in use here is v1.2, based on <a class="reference external" href="https://alkesgroup.broadinstitute.org/BOLT-LMM/BOLT-LMM_manual.html">bolt-lmm</a> documentation and other software support suggesting that
that’s the most efficient version accepted by all current tools. this may need to be changed in the future</p></li>
<li><p>the <a class="reference external" href="https://www.well.ox.ac.uk/~gav/bgen_format/">bgen</a> reformatting process is conducted using <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a>. the resultant <strong>*.sample</strong> files are slightly malformatted,
and so an additional step fixes the included <strong>NA</strong> values by setting them to <strong>0</strong>. this could well be changed in
a future version depending on upstream behavior</p></li>
</ul>
</li>
</ul>
<div class="topic">
<p class="topic-title">Debugging</p>
<p><a class="reference external" href="https://www.well.ox.ac.uk/~gav/bgen_format/">bgen</a> support in conversion tools is pretty limited. I’ve ended up using <a class="reference external" href="https://www.cog-genomics.org/plink/">plink</a> for VCF-&gt;BGEN conversion in two steps,
even bearing in mind the apparent bug in output <strong>*.sample</strong> format files created with it. But I could very much see
the possibility of needing a different adapter program in the future depending on one’s needs and any format discrepancies
I’ve not found, and it would have the benefit of potentially removing an extra rule/intermediate file from this pipeline.</p>
</div>
</div>
<div class="section" id="genomes-reference-data">
<span id="genomes-pipeline"></span><h2>1000 Genomes Reference Data<a class="headerlink" href="#genomes-reference-data" title="Permalink to this headline">¶</a></h2>
<p>Certain pipelines (see below) use derived information from the <a class="reference external" href="https://www.internationalgenome.org">1000 Genomes Project</a>. This is (always?) split
by supercontinent.</p>
<ul class="simple">
<li><p>Usage: <strong>make 1KG_files</strong></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p>a functional internet connection</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>the 1000 Genomes files downloaded are frozen at a particular latest release according to the configuration information
in <strong>Makefile.config</strong>. that can obviously be changed if you want</p></li>
<li><p>most target installations should actually have some sort of copy of the 1000 Genomes data present already somewhere
on their filesystem; however, this pipeline is not designed to support that as-is. it should be pretty easy to modify
if you really want</p></li>
</ul>
</li>
</ul>
<div class="topic">
<p class="topic-title">Debugging</p>
<p>This step in the process was the source of one of the weirdest bugs I’ve found during this development process.
I don’t think it should ever come up again, but do run the <cite>1KG_files-check</cite></p>
</div>
</div>
<div class="section" id="ld-score-bolt-reference-files">
<span id="ldsc-pipeline"></span><h2>LD Score/BOLT Reference Files<a class="headerlink" href="#ld-score-bolt-reference-files" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://github.com/bulik/ldsc">ldsc</a> and <a class="reference external" href="https://alkesgroup.broadinstitute.org/BOLT-LMM/BOLT-LMM_manual.html">bolt-lmm</a> use data files derived from 1000 Genomes reference supercontinental data. Though some standard
reference data are provided in these packages (at least at time of initial download), that information only
covered European subjects, and more generalized data were/are needed.</p>
<ul class="simple">
<li><p>Usage: <strong>make ldsc</strong></p></li>
<li><p>Dependencies:</p>
<ul>
<li><p><a class="reference internal" href="#genomes-pipeline">1000 Genomes pipeline</a></p></li>
<li><p>1000 Genomes supercontinental definitions (included by default)</p></li>
</ul>
</li>
<li><p>Assumptions:</p>
<ul>
<li><p>the reference files included with <a class="reference external" href="https://alkesgroup.broadinstitute.org/BOLT-LMM/BOLT-LMM_manual.html">bolt-lmm</a> have a description of (in theory) how they were generated, but when you
actually run <a class="reference external" href="https://github.com/bulik/ldsc">ldsc</a> you find that none of the output matches the included reference files exactly. this is probably
some sort of weird versioning issue. regardless, this pipeline just hacks the result into submission. that results
in some discrepancies from the stock reference files, but there’s no indication of exactly which subjects/variants
were used for those files, so that’s not unexpected. basically: ymmv</p></li>
<li><p>default built-in files include an African American (<strong>AFRAMR</strong>) meta-group for appropriate subjects. note however
that “African American” as a human genetics group label is a very heterogeneous group, so there’s no guarantee
that this reference group will be appropriate for a given set of African American study subjects</p></li>
</ul>
</li>
</ul>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">plco-analysis</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Preprocessing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#relatedness">Relatedness</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ancestry-estimation">Ancestry Estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genotype-cleaning-by-ancestry">Genotype Cleaning by Ancestry</a></li>
<li class="toctree-l2"><a class="reference internal" href="#imputed-data-reformatting">Imputed Data Reformatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genomes-reference-data">1000 Genomes Reference Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ld-score-bolt-reference-files">LD Score/BOLT Reference Files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Primary%20Analysis.html">Primary Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="Meta-Analysis.html">Meta-Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="Postprocessing.html">Postprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="Globus%20Distribution.html">Globus Distribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="User%20Vignette.html">User Vignette</a></li>
<li class="toctree-l1"><a class="reference internal" href="Advanced%20Development%20Notes.html">Advanced Development Notes</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="Installation.html" title="previous chapter">Installation</a></li>
      <li>Next: <a href="Primary%20Analysis.html" title="next chapter">Primary Analysis</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Cameron Palmer.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.4.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/Preprocessing.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>